### PR TITLE
Trim DataTypeImpl binary size

### PR DESCRIPTION
**Pull request by edgchen1**
**Description**
Update DataTypeImpl to reduce binary size.
- Change some functions to not be virtual.
- Refactor some helpers.

Modest decrease measured in SizeBench:
image

**Motivation and Context**
Binary size.

### CODE REVIEW

File: include/onnxruntime/core/framework/data_types.h [hunk 1]
```diff
@@ -914,18 +886,19 @@ class SequenceTensorType : public SequenceTensorTypeBase {
...
+    SequenceTypeHelper::Set(SequenceTypeHelper::GetElemType<TensorElemType>()->GetTypeProto(),
+                            MutableTypeProto());
   }
 };
 
 /**
  * \brief OpaqueType
  *
- * \param T - cpp runtume that implements the Opaque type
+ * \tparam T - cpp runtume that implements the Opaque type
```
Comment (yuslepukhin) [inline include/onnxruntime/core/framework/data_types.h:897]: >\ [](http://example.com/codeflow?start=3&length=1)

Why \tparam?

**Review by yuslepukhin — APPROVED** [review]
:shipit:

**Comment by hariharans29** [comment]
Was the 4KB decrease measured for Windows RelWithDebInfo by any chance ?  The minimal build reduction seems to be at 25KB!

**Comment by edgchen1** [comment]
That was a nice surprise. My configuration was Windows MinSizeRel - but next time I should probably measure other configurations/platforms too.

**Comment by hariharans29** [comment]
I had the minimal build binary size tracker chart opened on a tab and I noticed a dip and I tracked it down to this commit. Pretty cool gains :)

**Pull request merged**